shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -55,63 +85,103 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
-        }
-      }
-      if (errorMessages.size() != 0) {
-        throw new IOException(Joiner.on('\n').join(errorMessages));
-      }
-    } else if (status.isPersisted()) {
-      System.out.println(filePath + " is already persisted");
-    } else {
-      try {
-        FileSystemUtils.persistFile(mFileSystem, filePath);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new RuntimeException(e);
-      } catch (TimeoutException e) {
-        throw new RuntimeException(e);
+    mTotalFilesToPersist = mToPersist.size();
+    System.out.println("Found " + mTotalFilesToPersist + " files to persist.");
+    ExecutorService service = Executors.newFixedThreadPool(parallelism);
+    List<Future<Void>> persistCalls = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      persistCalls.add(service.submit(new PersistCallable()));
+    }
+    try {
+      for (Future<Void> call : persistCalls) {
+        call.get();
       }
-      System.out.println("persisted file " + filePath + " with size " + status.getLength());
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
     }
+    return 0;
   }
 
   @Override
   public String getUsage() {
-    return "persist <path> [<path> ...]";
+    return "persist [--parallelism <#>] <path> [<path> ...]";
   }
 
   @Override
   public String getDescription() {
     return "Persists files or directories currently stored only in Alluxio to the "
         + "UnderFileSystem.";
   }
+
+  /**
+   * Clears the queue to persist and internal variables. This class is not designed to be used
+   * concurrently or from an existing incomplete state.
+   */
+  private void resetState() {
+    mToPersist.clear();
+    mFilesPersisted.set(0);
+    mTotalFilesToPersist = 0;
+  }
+
+  /**
+   * Thread that polls the persist queue and persists files.
+   */
+  private class PersistCallable implements Callable<Void> {
+    @Override
+    public Void call() throws Exception {
+      AlluxioURI toPersist = mToPersist.poll();
+      while (toPersist != null) {
+        try {
+          FileSystemUtils.persistFile(mFileSystem, toPersist);
+          // There is a slight chance of out of order output to the client.
+          // If this becomes a problem we can consider locking instead of an atomic int.
+          String progress =
+              "(" + mFilesPersisted.incrementAndGet() + "/" + mTotalFilesToPersist + ")";
+          System.out.println(progress + " Persisted file " + toPersist);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+          throw new RuntimeException(e);
+        } catch (Exception e) {
+          System.out.println("Failed to persist file " + toPersist);
```

Comment:
log the error?

Reply:
How do you feel about having the error logged in the user log and just having this message printed to stdout?

Reply:
sgtm, though it should probably go to stderr

Reply:
Updated, I will keep stdout for now since we are not consistent in using stderr. I've created: https://alluxio.atlassian.net/browse/ALLUXIO-3386